PERF: speed up concat on Series by skipping unnecessary DataFrame creation

qwhelan opened the pull request:
Removes an unnecessary `DataFrame` creation when dealing solely with `Series` objects, which reduces runtime of `concat`.

- [ ] xref #23362
- [x] tests passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### Performance Comparison
```python
import pandas as pd
import numpy as np
a = np.random.randint(2000, 2100, size=100000)  # larger n than #23362 
b = np.random.randint(2000, 2100, size=100000)

x = pd.core.arrays.period_array(a, freq='B')
y = pd.core.arrays.period_array(b, freq='B')

s = pd.Series(x)
t = pd.Series(y)
```

#### Baseline
```python
%timeit x._concat_same_type([x, y])
202 µs ± 10.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit pd.concat([s, t], ignore_index=True)
839 µs ± 32.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

#### After
```python
%timeit pd.concat([s, t], ignore_index=True)
396 µs ± 12.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
And the excised code itself:
```python
In [13]: pd.DataFrame()._get_axis_number(0)
Out[13]: 0

In [14]: %timeit pd.DataFrame()._get_axis_number(0)
Out[14]: 312 µs ± 23.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
So roughly 40% of the runtime was being spent mapping `axis=0 -> axis=0`

qwhelan committed:
PERF: speed up concat on Series by making _get_axis_number() a classmethod

jreback reviewed (changes requested):
so this actually affects perf? pls show a before and after

qwhelan commented:
@jreback I was waiting for `asv` to finish before posting benchmarks, but I've updated the initial post with some basic benchmarks as that's taking longer than anticipated.

topper-123 commented on pandas/core/reshape/concat.py:
```diff
@@ -320,7 +320,8 @@ def __init__(self, objs, axis=0, join='outer', join_axes=None,
         if isinstance(sample, Series):
-            axis = DataFrame()._get_axis_number(axis)
+            if not isinstance(axis, int):
+                axis = DataFrame()._get_axis_number(axis)
```
This is a very heavy operation for a simple issue as getting the axis number.

You should try to get it done without instantiation a dataframe. Something like ``pd.DataFrame._get_axis_number(pd.DataFrame, axis)`` should probably work and be very fast.

qwhelan replied:
@topper-123 Thanks, was looking for something like that. Just pushed that change.

qwhelan replied:
Unfortunately, seems this breaks on 2.7 - I'm going to leave as is for 2.7 and only make the improvement for 3

jreback reviewed (changes requested):
this change just obfuscates things for very very little gain here. This is a constant time operation that we use all over the codebase.

You can *try* to make this a ``@staticmethod`` instead if you really want to get this in.

TomAugspurger commented:
@qwhelan doing this just on Py3 also feels hacky. I think everything in `_get_axis_number` is static, so you can do

```diff
diff --git a/pandas/core/generic.py b/pandas/core/generic.py
index db10494f0..14edc1d7f 100644
--- a/pandas/core/generic.py
+++ b/pandas/core/generic.py
@@ -358,18 +358,19 @@ class NDFrame(PandasObject, SelectionMixin):
             d.update(kwargs)
             return cls(data, **d)
 
-    def _get_axis_number(self, axis):
-        axis = self._AXIS_ALIASES.get(axis, axis)
+    @classmethod
+    def _get_axis_number(cls, axis):
+        axis = cls._AXIS_ALIASES.get(axis, axis)
         if is_integer(axis):
-            if axis in self._AXIS_NAMES:
+            if axis in cls._AXIS_NAMES:
                 return axis
         else:
             try:
-                return self._AXIS_NUMBERS[axis]
+                return cls._AXIS_NUMBERS[axis]
             except KeyError:
                 pass
         raise ValueError('No axis named {0} for object type {1}'
-                         .format(axis, type(self)))
+                         .format(axis, cls))
 
     def _get_axis_name(self, axis):
         axis = self._AXIS_ALIASES.get(axis, axis)
```

Will you see if the same change can be made to

- `_get_axis_name`
- `_get_axis`
- `_get_block_manager_axis`

Then, will you ensure that we have basic tests for using all those methods on classes (as well as instances).

qwhelan commented:
@jreback Yeah, I'm aware this change is trivial but just using it as an excuse to get my dev setup fully working for some more substantial perf patches that will be coming soon.

@TomAugspurger Thanks, I'll check into it.

qwhelan commented:
@TomAugspurger I implemented the `@classmethod` change for all but `_get_axis()` (which is used to get an axis instance, so shouldn't be a `@classmethod`).

There's decent existing test coverage but I added an explicit test of class vs instance invocations for all valid possible calls for good measure.

TomAugspurger commented:
Thanks, LGTM, though you have a linting error: 
https://github.com/pandas-dev/pandas/pull/23404#issuecomment-433784500

qwhelan commented:
@TomAugspurger Thanks, resolved.

jreback commented on pandas/tests/generic/test_generic.py:
```diff
@@ -1019,3 +1019,17 @@ def test_pipe_panel(self):
             with pytest.raises(ValueError):
                 result = wp.pipe((f, 'y'), x=1, y=1)
+
+    @pytest.mark.filterwarnings("ignore:\\nPanel:FutureWarning")
```
this is ok, but pls parameterize over box (e.g. pd.Series, pd.DataFrame). ok to not test on Panel.

jreback reviewed (changes requested):
small comment, ping on green.

jreback commented:
lgtm. ping on green.

qwhelan commented:
@jreback All green, thanks!

TomAugspurger commented on pandas/tests/generic/test_generic.py:
```diff
@@ -1019,3 +1019,15 @@ def test_pipe_panel(self):
+    @pytest.mark.parametrize('box', [pd.Series, pd.DataFrame])
+    def test_axis_classmethods(self, box):
+        obj = box()
+        values = list(box._AXIS_NAMES.keys()) + \
```
What's the current status on backslash to escape long lines. Someone (@jbrockmendel?) put a lot of effort in to clean these up. Do we have an issue to make these a linting rule?

TomAugspurger replied:
@qwhelan would you mind converting these to use opening and closing parenthesis to wrap the long lines?

qwhelan replied:
Done

TomAugspurger commented:
```
In [4]: %timeit x._concat_same_type([x, y])
   ...:
99.8 µs ± 2.25 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [5]: %timeit pd.concat([s, t], ignore_index=True, copy=False)
   ...:
   ...:
194 µs ± 3.98 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

so we're down to 2x. Let me know if you're interested in pushing this further @qwhelan (in a separate PR).

jreback commented:
thanks @qwhelan

qwhelan commented:
@TomAugspurger Yeah, I'm looking at some related issues if you'd like to keep that issue open for now.
